packages/vite/src/node/server/middlewares/transform.ts

```diff
@@ -48,6 +48,13 @@ export function transformMiddleware(
   server: ViteDevServer,
 ): Connect.NextHandleFunction {
   // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+
+  // check if public dir is inside root dir
+  const publicDir = normalizePath(server.config.publicDir)
+  const rootDir = normalizePath(server.config.root)
```

Comment:
https://github.com/vitejs/vite/blob/ea6a7a6eb86ebc8ae5cf5aff88446a1b44fecec9/packages/vite/src/node/config.ts#L490-L492
https://github.com/vitejs/vite/blob/ea6a7a6eb86ebc8ae5cf5aff88446a1b44fecec9/packages/vite/src/node/config.ts#L738
`config.root` is already normalized, so we don't need to normalize it here.
I think we can normalize the publicDir when we resolve the config as well.
https://github.com/vitejs/vite/blob/ea6a7a6eb86ebc8ae5cf5aff88446a1b44fecec9/packages/vite/src/node/config.ts#L650-L656

Reply:
Done with `config.root`. About `publicDir`, we got hit by it not being normalized here https://github.com/vitejs/vite/pull/15317, so I think it is a good idea. But I was thinking we should do this in the next minor or major because some plugins may expect it to not be normalized. Hopefully there aren't many instances like that but better to play safe.

I think that all the directory paths in the resolved config should be normalized and without a trailing slash: `root`, `publicDir`, `outDir`. We should check what we are doing with `base`. There are a lot of places were we are doing a `path.posix.join` or similar when we could be directly concatenating the strings if we trust the resolved config.